common/ceph_json: dump bool using f->dump_bool()

tchaikov opened the pull request:
<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".

  - The Signed-off-by line in every git commit is important; see <span class="x x-first x-last">[Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/master/</span>SubmittingPatches.rst<span class="x x-first x-last">)</span>
-->

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

Fixes: https://tracker.ceph.com/issues/55189

idryomov reviewed (commented):
> encode_json() is used by rgw and rbd. the former uses JSONObj
defined in common/ceph_json.h to decode JSON-encoded structs.
and it does not differentiate bool from str when decoding a boolean
value despite that it could have check the "quoted" member variable
of JSONObj for validating the type of value. so we should be fine.

Are all of these structs internal to RGW or might some of them be parsed by external entities?  A quick grep shows that there are at least a couple of instances that are dumped by `radosgw-admin` CLI tool so a release note is likely needed.  @cbodley, please take a look.

> when it comes to rbd, it does not use encode_json() for encoding
boolean values at the time of writing, so we should be safe as
well.

This isn't quite true -- it is used in `ImageCacheState::dump` and that JSON gets persisted too.  But we are changing it in https://github.com/ceph/ceph/pull/45684 and persistent image cache in general is a new feature so from the RBD perspective this change seems OK.

cbodley commented:
@idryomov i agree that a release note is probably warranted here, as `radosgw-admin` and our associated `/admin` REST APIs are both heavy on json

rgw multisite relies on these admin APIs for communication across clusters. each cluster could be running a different ceph version, so we do need to make sure changes here are backward-compatible. but i'd expect the ceph-object-corpus/dencoder tests to catch any issues here - the important rgw types should all be covered there

tchaikov commented:
thank you guys for your review and insights.  will

- correct the commit message regarding the impact of this change
- prepare a note on this change in release note.

tchaikov commented:
@idryomov and @cbodley updated and repushed. could you please take another look?

anthonyeleven commented on PendingReleaseNotes:
```diff
@@ -3,3 +3,41 @@
+  are returned without quotes so they can be decoded as boolean values in JSON.
+  The same applies to the `is_truncated` field returned by `/subscriptions/<sub-name>`.
+* RGW's response of `Action=GetTopicAttributes&TopicArn=<topic-arn>` REST API now
+  returns `HasStoredSecret` and `Persistent` as boolen in the JSON string
```
Should `boolen` be \`\`boolean\`\` or \`\`bool\`\`?

tchaikov replied:
i will change it to boolean, but without the quotes. as boolean or bool is not a type defined in JSON's official document . see https://www.json.org/json-en.html . `true` and `false` just literals. they are not defined as two possible values of a certain type. boolean is just "invented" here to help user to understand the represented value.

anthonyeleven commented on PendingReleaseNotes:
```diff
@@ -3,3 +3,41 @@
 * RGW's default backend for `rgw_enable_ops_log` changed from RADOS to file.
   The default value of `rgw_ops_log_rados` is now false, and `rgw_ops_log_file_path`
   defaults to "/var/log/ceph/ops-log-$cluster-$name.log".
+* RGW's pubsub interface now returns boolean fields using bool. Before this change,
```
Should `bool` be \`\`bool\`\` ?

tchaikov replied:
in terms of explanation above, how about replacing "bool" with `"true"` or `"false"`. it's probably kind of pedantic though. but more correct, IMHO. what do you think?

anthonyeleven replied:
Agreed

dmick replied:
not sure about that; every language binding for JSON that I know of will convert 'true' or 'false' to a boolean of the appropriate state, and while JSON doesn't "define" it, the bindings do.  Also, they're more than just string literals; they're distinguished in the spec as one of three "literal name tokens" ("null" being the third).  I think they're as close to boolean as any representation without an actual type definition could possibly be.

Some online references actually call them booleans, too (although I agree, pedantically, incorrectly).  The reason isn't so much that there isn't a boolean; the reason is that more that there are no variables for which to restrict the possible values.

And Javascript has booleans.

anthonyeleven replied:
I raised my initial question because I'm not that familar with the language / API in question, so I wasn't sure if `bool` here is short for `boolean`, if it's a specific literal, etc.

anthonyeleven commented:
LGTM for docs

tchaikov commented:
change since last revision

- s/boolen/boolean/: i managed to revert the proposed change by Anthony. fixed it in the latest revision.

cbodley commented:
jenkins test api

tchaikov commented:
change since last version

- fixed the `/{admin_path}/metadata/user?key={userid}` rgw_client API: do not call `strtobool()` anymore -- we already have a `bool` now.

tchaikov commented:
jenkins test api
